src/overlap.rs

```diff
@@ -0,0 +1,88 @@
+use std::sync::Arc;
+
+use itertools::Itertools;
+
+use errors::*;
+use ir::*;
+use solve::solver::Solver;
+
+impl Program {
+    pub fn check_overlapping_impls(&self) -> Result<()> {
+        let mut solver = Solver::new(&Arc::new(self.environment()), 10);
+
+        // Create a vector of references to impl datums, sorted by trait ref
+        let impl_data = self.impl_data.values().sorted_by(|lhs, rhs| {
+            lhs.binders.value.trait_ref.trait_id.cmp(&rhs.binders.value.trait_ref.trait_id)
+        });
+
+        // Group impls by trait.
+        let impl_groupings = impl_data.into_iter().group_by(|impl_datum| {
+            impl_datum.binders.value.trait_ref.trait_id
+        });
+
+        for (trait_id, impls) in &impl_groupings {
+            let impls: Vec<&ImplDatum> = impls.collect();
+
+            // For each pair, check their overlap by generating an "intersection"
+            // goal. In this case, success is an error - it means that there is at
+            // least one type in the intersection of these two impls.
+            for (lhs, rhs) in impls.into_iter().tuple_combinations() {
+                match solver.solve_goal(intersection_of(lhs, rhs)) {
+                    Ok(_)   => {
+                        let trait_id = self.type_kinds.get(&trait_id).unwrap().name;
+                        Err(Error::from_kind(ErrorKind::OverlappingImpls(trait_id)))
+                    }
+                    Err(_)  => Ok(())
+                }?;
+            }
+        }
+
+        Ok(())
+    }
+}
+
+// The goal to test overlap.
+//
+// If this goal succeeds, these two impls overlap.
+fn intersection_of(lhs: &ImplDatum, rhs: &ImplDatum) -> Canonical<InEnvironment<Goal>> {
+    fn params(impl_datum: &ImplDatum) -> &[Parameter] {
+        &impl_datum.binders.value.trait_ref.parameters
+    }
+
+    debug_assert!(params(lhs).len() == params(rhs).len());
+
+    let lhs_len = lhs.binders.len();
+
+    // Join the two impls' binders together 
+    let mut binders = lhs.binders.binders.clone();
+    binders.extend(rhs.binders.binders.clone());
+
+    // Upshift the rhs variables in params to account for the joined binders
+    let lhs_params = params(lhs).iter().cloned();
+    let rhs_params = params(rhs).iter().map(|param| param.up_shift(lhs.binders.len()));
```

Comment:
Nit: you use `lhs.binders.len()` here but `lhs_len` below; made me think they were different